File: 프로그래머스/1/12906. 같은 숫자는 싫어/같은 숫자는 싫어.cpp
#include <vector>
#include <deque>
#include <iostream>

using namespace std;

vector<int> solution(vector<int> arr)
{
    vector<int> answer;
    
    deque<int> s;

    for (int i{ 0 }; i < arr.size(); i++) {
        if (s.empty()) {
            s.push_back(arr[i]);
        }
        else if (s.back() != arr[i]) {
            s.push_back(arr[i]);
        }
    }

    int size = s.size();

    for (int i{ 0 }; i < size; i++) {
        answer.push_back(s.front());
        s.pop_front();
    }

    return answer;
}

File: 프로그래머스/1/12906. 같은 숫자는 싫어/같은 숫자는 싫어.java
import java.util.*;

public class Solution {
    public List<Integer> solution(int []arr) {
        List<Integer> ls = new ArrayList<>();
        Deque<Integer> dq = new ArrayDeque<>();
        
        for(int a : arr) {
            if(dq.size()==0 || a != dq.getLast()) {
                dq.add(a);
            }
        }
        
        int size = dq.size();
        
        for(int i=0; i<size; i++) {
            ls.add(dq.getFirst());
            dq.poll();
        }
        
        return ls;
    }
}